runtime/jcl/common/thread.cpp

```diff
@@ -603,28 +603,111 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
 #endif /* JAVA_SPEC_VERSION >= 19 */
 
 #if JAVA_SPEC_VERSION >= 20
+/**
+ * Frame walk function to find the most recent scoped value binding in the stack.
+ *
+ * @param currentThread the current thread
+ * @param walkState the stack walk state
+ * @return J9_STACKWALK_KEEP_ITERATING or J9_STACKWALK_STOP_ITERATING
+ */
+static UDATA
+findScopedValueBindingsWalkFunction(J9VMThread *currentThread, J9StackWalkState *walkState)
+{
+	UDATA rc = J9_STACKWALK_KEEP_ITERATING;
+
+	if (NULL != walkState->userData1) {
+		J9JavaVM *vm = currentThread->javaVM;
+		J9Method *method = walkState->method;
+		J9Class *methodClass = J9_CLASS_FROM_METHOD(method);
+		J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+		J9UTF8 *methodName = J9ROMMETHOD_NAME(romMethod);
+		if (((J9VMJAVALANGTHREAD_OR_NULL(vm) == methodClass)
+			|| (J9VMJAVALANGVIRTUALTHREAD_OR_NULL(vm) == methodClass)
+			|| ((J9Class *)walkState->userData3 == methodClass))
+			&& J9UTF8_LITERAL_EQUALS(J9UTF8_DATA(methodName), J9UTF8_LENGTH(methodName), "runWith")
+		) {
+			rc = J9_STACKWALK_STOP_ITERATING;
+		} else if (0 == walkState->inlineDepth) {
+			walkState->userData1 = NULL;
+		}
+	}
+
+	return rc;
+}
+
+/**
+ * Object slot walk function to find the most recent scoped value binding in the object slots.
+ *
+ * @param currentThread the current thread
+ * @param walkState the stack walk state
+ * @param slot pointer to slot containing an object pointer
+ * @param stackLocation pointer to the slot in the stack containing possibly compressed object pointer
+ */
+static void
+findScopedValueBindingOSlotWalkFunction(J9VMThread *currentThread, J9StackWalkState *walkState, j9object_t *slot, const void *stackLocation)
+{
+	j9object_t slotObject = *slot;
+	J9Class *clazz = J9OBJECT_CLAZZ(currentThread, slotObject);
+	J9Class *snapshotClazz = (J9Class *)walkState->userData2;
+
+	if (isSameOrSuperClassOf(snapshotClazz, clazz)) {
+		if (NULL == walkState->userData1) {
+			walkState->userData1 = (void *)slotObject;
+		} else {
+			Assert_JCL_true(0);
```

Comment:
This would read better with the assertion first, then the assignment to the slot (with some comments please).